File: Geometry/3d_point.cpp
template <class T>
struct Point {
    T x, y, z;

    /* O(1) - Scalar sum. */
    Point<T> operator+(const T &c) const {
        return {this->x + c, this->y + c, this->z + c};
    }

    /* O(1) - Scalar subtraction. */
    Point<T> operator-(const T &c) const {
        return {this->x - c, this->y - c, this->z - c};
    }

    /* O(1) - Scalar multiplication. */
    Point<T> operator*(const T &c) const {
        return {this->x * c, this->y * c, this->z * c};
    }

    /* O(1) - Scalar division. */
    Point<T> operator/(const T &c) const {
        return {this->x / c, this->y / c, this->z / c};
    }

    /* O(1) - Negative vector. */
    Point<T> operator-() const {
        return {-this->x, -this->y, -this->z};
    }

    /* O(1) - Vector sum. */
    Point<T> operator+(const Point<T> &b) const {
        return {this->x + b.x, this->y + b.y, this->z + b.z};
    }

    /* O(1) - Vector subtraction. */
    Point<T> operator-(const Point<T> &b) const {
        return {this->x - b.x, this->y - b.y, this->z - b.z};
    }

    /* O(1) - Dot product. */
    T operator*(const Point<T> &b) const {
        return (this->x * b.x) + (this->y * b.y) + (this->z * b.z);
    }

    /* O(1) - Cross product. */
    Point<T> operator^(const Point<T> &b) const {
        return {(this->y * b.z) - (this->z * b.y), (this->z * b.x) - (this->x * b.z), (this->x * b.y) - (this->y * b.x)};
    }

    /* O(1) - Squared norm. */
    T operator!() const {
        return (this->x * this->x) + (this->y * this->y) + (this->z * this->z);
    }

    /* O(1) - Lesser than. */
    bool operator<(const Point<T> &b) const {
        if (this->x == b.x) {
            if (this->y == b.y) {
                return this->z < b.z;
            }

            return this->y < b.y;
        }

        return this->x < b.x;
    }

    /* O(1) - Greater than. */
    bool operator>(const Point<T> &b) const {
        return b < *this;
    }

    /* O(1) - Equal. */
    bool operator==(const Point<T> &b) const {
        return !(*this < b) and !(*this > b);
    }

    /* O(1) - Different. */
    bool operator!=(const Point<T> &b) const {
        return !(*this == b);
    }
};

/* O(1) - String conversion. */
template <class T>
string to_string(const Point<T> &p) {
    return "(" + to_string(p.x) + ", " + to_string(p.y) + +", " + to_string(p.z) + ")";
}

File: Geometry/3d_point.py
class Point:
	# O(1) - Constructor.
	def __init__(self, x, y, z):
		self.x, self.y, self.z = x, y, z

	# O(1) - Vector sum.
	def __add__(self, p):
		return Point(self.x + p.x, self.y + p.y, self.z + p.z)

	# O(1) - Negative vector.
	def __neg__(self):
		return Point(-self.x, -self.y, -self.z)

	# O(1) - Vector subtraction.
	def __sub__(self, p):
		return Point(self.x - p.x, self.y - p.y, self.z - p.z)

	# O(1) - Dot product.
	def __mul__(self, p):
		return self.x * p.x + self.y * p.y + self.z * p.z

	# O(1) - Cross product.
	def __xor__(self, p):
		return Point(self.y * p.z - self.z * p.y, self.z * p.x - self.x * p.z, self.x * p.y - self.y * p.x)

	# O(1) - Lesser than.
	def __lt__(self, p):
		if self.x == p.x:
			if self.y == p.y:
				return self.z < p.z

			return self.y < p.y

		return self.x < p.x

	# O(1) - Greater than.
	def __gt__(self, p):
		return p < self

	# O(1) - Equal.
	def __eq__(self, p):
		return self.x == p.x and self.y == p.y and self.z == p.z

	# O(1) - Different.
	def __ne__(self, p):
		return not self == p

	# O(1) - String conversion.
	def __str__(self):
		return "(%s, %s, %s)" % (str(self.x), str(self.y), str(self.z))

	# O(1) - String conversion for printing arrays.
	def __repr__(self):
		return str(self)